### PR TITLE
Boundary Condition : Interface with existing PML

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
This PR interfaces the new boundary input, `field.boundary_lo = PML` and `field.boundary_hi = PML` with the pml implementation the code, by setting `do_pml`, `do_pml_Lo` and `do_pml_Hi` accordingly.

Additionally, if `num_levels > 0`, then `do_pml = 1` and `do_pml_Lo_MR` and `do_pml_Hi_MR` are set to 1 in each direction.

However, if the fine patch co-incides with the domain boundary, then the domain boundary condition will be applied and the pml flag for the level1 and higher patches will be set to the value of the pml flag consistent with the domain boundary.
That is, if the domain boundary is pml, then the pml flags for MR patches will be set to 1 in that direction.
If domain boundary is pec, periodic, or pmc, then the pml flag for the MR patches will be set to 0.

I have tested this with the test in `Examples/Tests/particles_in_PML/` using the input parameters
```
boundary.field_lo = pml pml pml
boundary.field_hi = pml pml pml
```
The particles in this test-case are still treated based on the input parameters, `warpx.do_pml_j_damping=1`

Note that, `geometric.is_periodic` and `warpx.do_pml` are not used, since the field boundary condition will override those parameters.

[inputs_mr_3d_BC.txt](https://github.com/ECP-WarpX/WarpX/files/6224879/inputs_mr_3d_BC.txt)

Screenshot from 2021-03-29 16-43-46

To be merged after PR #1730

### CODE REVIEW

File: Source/Utils/WarpXUtil.cpp [hunk 1]
```diff
@@ -408,6 +408,7 @@ void ReadBCParams ()
             }
 
         }
+
```
Comment (RevathiJambunathan) [inline Source/Utils/WarpXUtil.cpp]:
```suggestion
```

File: Docs/source/usage/parameters.rst [hunk 2]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
 --------------------------
 
 * ``boundary.field_lo`` and ``boundary_field_hi`` (`2 strings` for 2D, `3 strings` for 3D)
-    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for `geometry.is_periodic` will be set, overriding the user-input for the input parameter, `geometry.is_periodic`.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst]: Just the odd way of `.rst` to verbatim :sweat_smile: 
```suggestion
    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for ``geometry.is_periodic`` will be set, overriding the user-input for the input parameter, ``geometry.is_periodic``.
```

File: Docs/source/usage/parameters.rst [hunk 3]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
...
     Options are:
     * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
+    * ``PML``: This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst]: I was wondering from the example in the PR description: Are these options lower case/upper case or case insensitive?

File: Docs/source/usage/parameters.rst [hunk 4]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
...
+    Additional pml algorithms can be explored using the parameters `warpx.do_pml_in_domain`, `warpx.do_particles_in_pml`, and `warpx.do_pml_j_damping`.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst]:
```suggestion
    Additional pml algorithms can be explored using the parameters ``warpx.do_pml_in_domain``, ``warpx.do_particles_in_pml``, and ``warpx.do_pml_j_damping``.
```

File: Docs/source/usage/parameters.rst (hunk 3 above)
Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst]: it is case insensitive. should I change this in the doc, so that users dont think it is all upper case?

File: Docs/source/usage/parameters.rst [hunk 5]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
 --------------------------
 
 * ``boundary.field_lo`` and ``boundary_field_hi`` (`2 strings` for 2D, `3 strings` for 3D)
-    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for ``geometry.is_periodic`` will be set, overriding the user-input for the input parameter, ``geometry.is_periodic``.
     Options are:
     * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
+    * ``PML``: This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```
Comment (RevathiJambunathan) [inline Docs/source/usage/parameters.rst]:
```suggestion
    * ``pml``: This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```

File: Docs/source/usage/parameters.rst (hunk 3 above)
Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst]: just changed it to lower case, to avoid users having to capitalize

**Comment by EZoni** [comment]
@RevathiJambunathan I will review the PR more thoroughly soon, but I think we can try something to fix the CI tests that are currently failing. They were not setting `warpx.do_pml` explicitly in their input files, and since you changed the default value of `WarpX::do_pml` set in Source/WarpX.H (from `True` to `False`), I think you would need to add `warpx.do_pml=1` explicitly among the `runtime_params` of the failing tests in Source/Regression/WarpX-tests.ini, in order to recover the old behavior. This fixes for sure the test `PlasmaMirror`, I think it will fix the others too (but I have some problems running them locally, hence I couldn't check).

**Comment by RevathiJambunathan** [comment]
@EZoni that is aboslutely right!
This is the change I wanted to test locally first using the run_script.sh file before I make the commit and push it.
Thanks for sharing with me how to test this locally, offline :)

**Comment by EZoni** [comment]
@RevathiJambunathan Great! Let's see if it fixes all the tests. Note that in order to run locally the test `LaserIonAcc2d`, which requires openPMD, you will need to make the additional modification implemented in #1852 (otherwise the executable for that test is built without openPMD and the test appears to crash just because of that). I noticed only today thanks to your PR here, while checking the tests that were failing.

File: Docs/source/usage/parameters.rst [hunk 6]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
 --------------------------
 
 * ``boundary.field_lo`` and ``boundary_field_hi`` (`2 strings` for 2D, `3 strings` for 3D)
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst:185]: I am wondering with `warpx.do_pml` now from the implicit on to off value change, if the default value for these options should be pml?

Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst:185]: When we had discussed this offline, during the meeting, we had thought that the default should be PEC. (PR in progress for this capability )

Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst:185]: But, it would be good to revisit, because `do_pml` was always 1 until now

Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst:185]: Thank you @ax3l for the offline discussion. We are setting default field boundary to pml to ensure that the default behaviour remains the same before and after the PR.

File: Docs/source/usage/parameters.rst [hunk 7]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
...
-    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for ``geometry.is_periodic`` will be set, overriding the user-input for the input parameter, ``geometry.is_periodic``. If not set, the default value for the fields at the domain boundary will be set to pml.
     Options are:
     * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
+    * ``pml``: This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst]: Let's also document the default in the docs, e.g.:
```suggestion
    * ``pml`` (default): This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```

Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst]: Thank you @ax3l . I had added a statement above "If not set, the default value for the fields at the domain boundary will be set to pml.".
But I agree that it is good to repeat what is the default in front of the option.

**Comment by RemiLehe** [comment]
The last commit seems to have broken the automated tests. @RevathiJambunathan Would you have time to look into this?

**Comment by RevathiJambunathan** [comment]
Oh I dont think the automated tests like default pml, for cases where we need periodic boundary.

**Comment by ax3l** [comment]
@dpgrote can you help us figuring out what we need to change in `Python_Langmuir_rz_multimode` (`Examples/Tests/Langmuir/PICMI_inputs_langmuir_rz_multimode_analyze.py`)?

I see the boundaries for this RZ test are:
```py
grid = picmi.CylindricalGrid(...
                             lower_boundary_conditions = ['dirichlet', 'periodic'],
                             upper_boundary_conditions = ['dirichlet', 'periodic'],
                             ...
)
```

and this PR removes the `do_pml = 1` default in lieu of new options `boundary.field_lo/hi` that are defaulting to PML. I think this might need an update in the PICMI logic for boundaries?

File: Examples/Modules/gaussian_beam/PICMI_inputs_gaussian_beam.py [hunk 8]
```diff
@@ -31,6 +31,7 @@
 
 solver = picmi.ElectromagneticSolver(grid = grid,
                                      cfl = 1.,
+                                     warpx_do_pml = True,
```
Comment (RevathiJambunathan) [inline Examples/Modules/gaussian_beam/PICMI_inputs_gaussian_beam.py]:
```suggestion
```

File: Examples/Physics_applications/laser_ion/inputs [hunk 9]
```diff
@@ -25,7 +25,8 @@ amr.n_cell = 2688 3712
 amr.max_level = 0
 geometry.prob_lo = -7.5e-6 -5.e-6
 geometry.prob_hi =  7.5e-6 25.e-6
-geometry.is_periodic = 0 0 0  # non-periodic (default)
+geometry.is_periodic = 0 0  # non-periodic (default)
+warpx.do_pml = 1
```
Comment (RevathiJambunathan) [inline Examples/Physics_applications/laser_ion/inputs]:
```suggestion
```

File: Examples/Physics_applications/plasma_acceleration/PICMI_inputs_plasma_acceleration.py [hunk 10]
```diff
@@ -28,7 +28,7 @@
                              moving_window_velocity = moving_window_velocity,
                              warpx_max_grid_size=32)
 
-solver = picmi.ElectromagneticSolver(grid=grid, cfl=1)
+solver = picmi.ElectromagneticSolver(grid=grid, cfl=1, warpx_do_pml = True)
```
Comment (RevathiJambunathan) [inline Examples/Physics_applications/plasma_acceleration/PICMI_inputs_plasma_acceleration.py]:
```suggestion
solver = picmi.ElectromagneticSolver(grid=grid, cfl=1)
```

File: Examples/Physics_applications/plasma_mirror/inputs_2d [hunk 11]
```diff
@@ -28,6 +28,7 @@ my_constants.zc2   = 24.05545177444479562e-6
 warpx.cfl = 1.0
 warpx.use_filter = 1
 algo.load_balance_intervals = 66
+warpx.do_pml = 1
```
Comment (RevathiJambunathan) [inline Examples/Physics_applications/plasma_mirror/inputs_2d]:
```suggestion
```

File: Docs/source/usage/parameters.rst [hunk 12]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
...
-    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for ``geometry.is_periodic`` will be set, overriding the user-input for the input parameter, ``geometry.is_periodic``. If not set, the default value for the fields at the domain boundary will be set to pml.
     Options are:
     * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
+    * ``pml`` (default): This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for `do_pml`. See the :ref:`PML theory section <theory-bc>` for more details.
```
Comment (EZoni) [inline Docs/source/usage/parameters.rst]: Would it be OK to also mention the prefix `warpx` when you mention the input parameter `do_pml`? Or is this meant to be fully removed soon anyways?
```suggestion
    * ``pml`` (default): This option can be used to add Perfectly Matched Layers (PML) around the simulation domain. It will override the user-defined value provided for ``warpx.do_pml``. See the :ref:`PML theory section <theory-bc>` for more details.
```

File: Source/BoundaryConditions/PML.cpp [hunk 13]
```diff
@@ -586,17 +583,15 @@ PML::PML (const int lev, const BoxArray& grid_ba, const DistributionMapping& /*g
...
-                    domain1.growLo(idim, -ncell/ref_ratio[idim]);
-                }
-                if (do_pml_Hi[idim]){
-                    // ncell is divided by refinement ratio to ensure that the
-                    // physical width of the PML region is equal is fine and coarse patch
-                    domain1.growHi(idim, -ncell/ref_ratio[idim]);
-                }
+            if (do_pml_Lo[idim]){
+                // ncell is divided by refinement ratio to ensure that the
+                // physical width of the PML region is equal is fine and coarse patch
```
Comment (EZoni) [inline Source/BoundaryConditions/PML.cpp]: This comment was already there before your PR, but since you are modifying these lines anyways, would you mind fixing it as well?
```suggestion
                // physical width of the PML region is equal in fine and coarse patch
```

File: Source/BoundaryConditions/PML.cpp [hunk 14]
```diff
@@ -586,17 +583,15 @@ PML::PML (const int lev, const BoxArray& grid_ba, const DistributionMapping& /*g
...
+                domain1.growLo(idim, -ncell/ref_ratio[idim]);
+            }
+            if (do_pml_Hi[idim]){
+                // ncell is divided by refinement ratio to ensure that the
+                // physical width of the PML region is equal is fine and coarse patch
```
Comment (EZoni) [inline Source/BoundaryConditions/PML.cpp]: This comment was already there before your PR, but since you are modifying these lines anyways, would you mind fixing it as well?
```suggestion
                // physical width of the PML region is equal in fine and coarse patch
```

**Review by EZoni — APPROVED** [review]
Thank you for this PR, Reva! I see that all the CI tests are successful, so all the various defaults seem correct and consistent now. The changes look good to me, I just left a couple of very minor comments.

File: Docs/source/usage/parameters.rst (hunk 12 above)
Comment (RevathiJambunathan) [reply Docs/source/usage/parameters.rst]: it is meant to be removed anyway. but good to be consistent :) Thank you!

File: Docs/source/usage/parameters.rst [hunk 15]
```diff
@@ -183,9 +183,11 @@ Domain Boundary Conditions
 --------------------------
 
 * ``boundary.field_lo`` and ``boundary_field_hi`` (`2 strings` for 2D, `3 strings` for 3D)
-    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Boundary conditions applied to field at the lower and upper domain boundaries. Depending on the type of boundary condition, the value for ``geometry.is_periodic`` will be set, overriding the user-input for the input parameter, ``geometry.is_periodic``. If not set, the default value for the fields at the domain boundary will be set to pml.
     Options are:
     * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst:189]: This is an odd thing with `.rst` lists: this needs an empty newline before the list starts to render correctly (see GitHub "preview"):
```suggestion

    * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input file, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
```

**Review by ax3l — APPROVED** [review]
Thank you!! I'll push a minor fix for the docs and then it goes in :tada:
